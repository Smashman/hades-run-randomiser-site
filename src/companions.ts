import { getRandomItemFromArray, Item, Options } from './utils';
import { ItemList } from './utils';

export class Companion extends Item {
    level: number = 1;
    isMaxCodex: boolean = false;
    constructor(public name: string, public giver: string) {
        super();
    }

    static maxLevel: number = 5;

    get isMaxLevel(): boolean {
        return this.level === Companion.maxLevel;
    }
}

export class Companions extends ItemList<Companion> {
    getRandomCompanion(companionOptions?: CompanionOptions): Companion | null {
        const companions = (companionOptions?.ignoreMaxCodex && !this.isAllUnlockedMaxLevel) ? [...this.unlocked.filter(companion => !companion.isMaxCodex)] : [...this.unlocked];
        return getRandomItemFromArray(companions) ?? null;
    }

    maxCodexAll() {
        this.items.forEach(companion => companion.isMaxCodex = true);
    }

    get isAllMaxCodex() {
        return this.items.every(companion => companion.isMaxCodex);
    }

    maxLevelAll() {
        this.items.forEach(companion => companion.level = Companion.maxLevel);
    }

    get isAllUnlockedMaxLevel() {
        return this.unlocked.every(companion => companion.isMaxLevel);
    }

    get isAllMaxLevel() {
        return this.items.every(companion => companion.isMaxLevel);
    }
}

export interface CompanionOptions extends Options {
    ignoreMaxCodex?: boolean;
}