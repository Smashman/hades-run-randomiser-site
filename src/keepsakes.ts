import { getRandomItemFromArray, Options, ItemList, Item, Level } from './utils';

export interface KeepsakeOptions extends Options {
    ignoreMaxLevel?: boolean;
    changeBetweenBiomes?: boolean;
}

export class Keepsakes extends ItemList<Keepsake> {
    getRandomKeepsakes(keepsakeOptions?: KeepsakeOptions): KeepsakeConfiguration {
        const keepsakes = (keepsakeOptions?.ignoreMaxLevel && !this.isAllUnlockedMaxLevel) ? [...this.unlocked.filter(keepsake => !keepsake.isMaxLevel)] : [...this.unlocked];

        if (keepsakes.length === 0) {
            return null;
        }

        let keepsakesChosen: Keepsake[] = [getRandomItemFromArray(keepsakes)];

        if (keepsakeOptions?.changeBetweenBiomes) {
            for (let i = 1; i <= 3; i++) {
                keepsakesChosen = [...keepsakesChosen, getRandomItemFromArray(keepsakes.filter(keepsake => !keepsakesChosen.includes(keepsake))) ?? null];
            }
        }

        return keepsakesChosen as KeepsakeConfiguration;
    }

    maxLevelAll() {
        this.items.forEach(keepsake => keepsake.level.setMaxLevel());
    }

    get isAllUnlockedMaxLevel() {
        return this.unlocked.every(keepsake => keepsake.isMaxLevel);
    }

    get isAllMaxLevel() {
        return this.items.every(keepsake => keepsake.isMaxLevel);
    }
}

export class Keepsake extends Item {
    level: Level = new Level(1, Keepsake.minLevel, Keepsake.maxLevel);
    constructor(public name: string, public giver: string, public icon: string, public isHidden: boolean = false) {
        super();
    }

    static minLevel = 1;
    static maxLevel = 3;

    get isMaxLevel(): boolean {
        return this.level.isMaxLevel();
    }
}

export type KeepsakeConfiguration = [Keepsake] | [Keepsake, Keepsake, Keepsake, Keepsake] | null;