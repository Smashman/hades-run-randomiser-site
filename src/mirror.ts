import { getRandomItemFromArray, Item, Options, StorableItemData } from './utils';

export interface MirrorOptions extends Options {
};

export const defaultMirrorOptions: MirrorOptions = {
    enabled: true,
};

export class Mirror {
    groups: MirrorGroup[] = [];

    addGroup(sets: [[MirrorTalent, MirrorTalent], [MirrorTalent, MirrorTalent]], isUnlocked: boolean = false) {
        const mirrorTalentSets = <[MirrorTalentSet, MirrorTalentSet]>sets.map(set => new MirrorTalentSet(set));
        this.groups = [...this.groups, new MirrorGroup(mirrorTalentSets, isUnlocked)];
        return this;
    }

    get unlockedGroups() {
        return this.groups.filter(group => group.isUnlocked);
    }

    unlockAll() {
        this.groups.forEach(group => group.unlock());
    }

    getRandom(): MirrorConfiguration {
        return new MirrorConfiguration(this.unlockedGroups.map(group => group.getRandom()).reduce((acc, val) => [...acc, ...val], <MirrorTalent[]>[]));
    }

    toStorableData() {
        return {
            groups: this.groups.map(group => group.toStorableData())
        }
    }
}

export class MirrorGroup extends Item{
    constructor(public sets: [MirrorTalentSet, MirrorTalentSet], public isUnlocked: boolean = false) {
        super();
    }

    getRandom() {
        return <[MirrorTalent, MirrorTalent]>this.sets.map(set => set.getRandom());
    }
}

export class MirrorTalentSet {
    constructor(public talents: [MirrorTalent, MirrorTalent]) {
        talents[0].colour = "red";
        talents[1].colour = "green";
    }

    getRandom(): MirrorTalent {
        return getRandomItemFromArray(this.talents);
    }
}

export class MirrorTalent {
    rankCount: number;
    colour: "red" | "green";
    totalCost: number;

    constructor(public name: string, public rankCosts: number[]) {
        this.rankCount = rankCosts.length;
        this.totalCost = rankCosts.reduce((totalCost, cost) => totalCost + cost);
    }
}

export class MirrorConfiguration {
    totalDarkness: number;
    constructor(public talents: MirrorTalent[]){
        this.totalDarkness = talents.reduce((totalCost, talent) => totalCost + talent.totalCost, 0);
    };
}