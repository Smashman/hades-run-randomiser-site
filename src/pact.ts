import { getRandomItemFromArray, Options } from './utils';

export interface HeatRange {
    min?: number;
    max?: number;
}

export interface PactOptions extends Options {
    heatRange?: HeatRange
}

export class Pact {
    conditions: PactCondition[] = [];

    addCondition(pactCondition: PactCondition) {
        this.conditions = [...this.conditions, pactCondition];
        return this;
    }

    getRandomWithRange(heatRange?: HeatRange) {
        return PactConfiguration.generateFromHeatRange(this, heatRange);
    }

    unlockAll() {
        this.conditions.forEach(condition => condition.enforceLimit = false);
    }

    get maxHeat(): number {
        return this.conditions.reduce((maxTotal, condition) => maxTotal + condition.totalHeat, 0);
    }

    get initialHeat(): number {
        return this.conditions.reduce((total, condition) => total + condition.initialHeat, 0);
    }
}

export class PactCondition {
    enforceLimit: boolean = false;
    private _initialRank: number = 0;
    constructor(public name: string, public readonly rankCosts: number[], public rankLimit?: number) {
        this.enforceLimit = !!rankLimit;
    }

    get initialRank(): number {
        return this._initialRank;
    }

    set initialRank(rank: number) {
        this._initialRank = rank > this.rankCount ? this.rankCount : rank;
    }

    get rankCount(): number {
        return this.rankLimit && this.enforceLimit ? this.rankLimit : this.rankCosts.length;
    }

    get totalHeat(): number {
        return (this.rankLimit && this.enforceLimit ? this.rankCosts.slice(0, this.rankLimit) : this.rankCosts).reduce((totalHeat, cost) => totalHeat + cost);
    }

    get initialHeat(): number {
        return this.rankCosts.slice(0, this.initialRank).reduce((total, value) => total + value, 0);
    }
}

export class PactConfiguration {
    conditionConfigurations: PactConditionConfiguration[];
    heatTotal: number = 0;
    heatTarget: number = 0;
    heatRange: HeatRange;

    constructor(pact: Pact, heatRange?: HeatRange, ranks?: number[]) {
        this.conditionConfigurations = pact.conditions.map((condition, index) => new PactConditionConfiguration(condition, (ranks && ranks[index]) || null));
        this.heatRange = {
            min: heatRange?.min || pact.initialHeat,
            max: heatRange?.max || pact.maxHeat,
        };
    }

    static generateFromHeatRange(pact: Pact, heatRange?: HeatRange) {
        const pactConfiguration = new PactConfiguration(pact, heatRange);

        const configHeatRange = pactConfiguration.heatRange;


        const target = Math.floor(Math.random() * (configHeatRange.max - configHeatRange.min + 1) + configHeatRange.min);

        let current = pact.initialHeat;

        let remainingConditions = [...pactConfiguration.conditionConfigurations.filter((conditionConfig) => conditionConfig.rank !== conditionConfig.condition.rankCount)];


        while (current < target) {
            const randomCondition = getRandomItemFromArray(remainingConditions);
            current += randomCondition.rankUp();

            if (randomCondition.rank === randomCondition.condition.rankCount) {
                remainingConditions = [...remainingConditions.filter(condition => randomCondition !== condition)];
            }
        }

        pactConfiguration.heatTotal = current;
        pactConfiguration.heatTarget = target;

        return pactConfiguration;
    }
}

export class PactConditionConfiguration {
    private _rank: number = 0;

    constructor(public condition: PactCondition, rank?: number) {
        this.rank = rank || condition.initialRank;
    }

    get heat(): number {
        return this.condition.rankCosts.slice(0, this.rank).reduce((total, value) => total + value, 0);
    }

    get rank(): number {
        return this._rank;
    }

    set rank(rank: number) {
        this._rank = rank > this.condition.rankCount ? this.condition.rankCount : rank;
    }

    rankUp(): number {
        if (this.rank < this.condition.rankCount) {
            this.rank++;
            return this.condition.rankCosts[this.rank-1];
        }
        else {
            return null;
        }
    }
}