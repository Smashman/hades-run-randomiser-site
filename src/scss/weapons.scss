@import './style.scss';

.weapons {
    display: flex;
    flex-wrap: wrap;
}

.weapon, .aspect {
    .weapon-name, .aspect-name {
        font-family: $spectral-sc;
        font-weight: bold;
    }
    .weapon-title {
        font-family: $crimson-text;
        font-style: italic;
    }
}

.weapon {
    border: 1px black solid;
    width: 300px;
    .weapon-name {
        font-size: 1.3em;
    }
}

.aspect {
    border: 1px blue solid;
    height: 90px;
    .aspect-name {
        font-family: $alegreya-sans-sc;
    }
}