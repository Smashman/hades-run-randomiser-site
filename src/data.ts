import { Weapon, WeaponAspect, Weapons } from './weapons';
import { Keepsake, Keepsakes } from './keepsakes';
import { Companion, Companions } from './companions';
import { Mirror, MirrorTalent } from './mirror';
import { Pact, PactCondition } from './pact';
import { fillArray } from './utils';
import * as React from 'react';
import * as keepsakeIcons from './img/keepsake';
import * as weaponIcons from './img/weapon';

export const weapons = new Weapons().addItems([
    new Weapon('Stygius', 'Blade of the Underworld', 'sword', true)
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.sword.zagreus, false, true),
        new WeaponAspect('Nemesis', weaponIcons.sword.nemesis),
        new WeaponAspect('Poseidon', weaponIcons.sword.poseidon),
        new WeaponAspect('Arthur', weaponIcons.sword.arthur, true),
    ]),
    new Weapon('Varatha', 'Eternal Spear', 'spear')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.spear.zagreus, false, true),
        new WeaponAspect('Achilles', weaponIcons.spear.achilles),
        new WeaponAspect('Hades', weaponIcons.spear.hades),
        new WeaponAspect('Guan Yu', weaponIcons.spear.guanYu, true),
    ]),
    new Weapon('Aegis', 'Shield of Chaos', 'shield')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.shield.zagreus, false, true),
        new WeaponAspect('Chaos', weaponIcons.shield.chaos),
        new WeaponAspect('Zeus', weaponIcons.shield.zeus),
        new WeaponAspect('Beowulf', weaponIcons.shield.beowulf, true),
    ]),
    new Weapon('Coronacht', 'Heart-Seeking Bow', 'bow')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.bow.zagreus, false, true),
        new WeaponAspect('Chiron', weaponIcons.bow.chiron),
        new WeaponAspect('Hera', weaponIcons.bow.hera),
        new WeaponAspect('Rama', weaponIcons.bow.rama, true),
    ]),
    new Weapon('Malphon', 'Twin Fists', 'fists')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.fists.zagreus, false, true),
        new WeaponAspect('Talos', weaponIcons.fists.talos),
        new WeaponAspect('Demeter', weaponIcons.fists.demeter),
        new WeaponAspect('Gilgamesh', weaponIcons.fists.gilgamesh, true),
    ]),
    new Weapon('Exagryph', 'Adamant Rail', 'gun')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.gun.zagreus, false, true),
        new WeaponAspect('Eris', weaponIcons.gun.eris),
        new WeaponAspect('Hestia', weaponIcons.gun.hestia),
        new WeaponAspect('Lucifer', weaponIcons.gun.lucifer, true),
    ]),
]);

export const keepsakes = new Keepsakes().addItems([
    new Keepsake('Old Spiked Collar', 'Cerberus', keepsakeIcons.oldSpikedCollar),
    new Keepsake('Myrmidon Bracer', 'Achilles', keepsakeIcons.myrmidonBracer),
    new Keepsake('Black Shawl', 'Nyx', keepsakeIcons.blackShawl),
    new Keepsake('Pierced Butterfly', 'Thanatos', keepsakeIcons.piercedButterfly),
    new Keepsake('Bone Hourglass', 'Charon', keepsakeIcons.boneHourglass),
    new Keepsake('Chthonic Coin Purse', 'Hypnos', keepsakeIcons.chthonicCoinPurse),
    new Keepsake('Skull Earring', 'Megaera', keepsakeIcons.skullEarring),
    new Keepsake('Distant Memory', 'Orpheus', keepsakeIcons.distantMemory),
    new Keepsake('Harpy Feather Duster', 'Dusa', keepsakeIcons.harpyFeatherDuster),
    new Keepsake('Lucky Tooth', 'Skelly', keepsakeIcons.luckyTooth),
    new Keepsake('Thunder Signet', 'Zeus', keepsakeIcons.thunderSignet),
    new Keepsake('Conch Shell', 'Poseidon', keepsakeIcons.conchShell),
    new Keepsake('Owl Pendant', 'Athena', keepsakeIcons.owlPendant),
    new Keepsake('Eternal Rose', 'Aphrodite', keepsakeIcons.eternalRose),
    new Keepsake('Blood-Filled Vial', 'Ares', keepsakeIcons.bloodFilledVial),
    new Keepsake('Adamant Arrowhead', 'Artemis', keepsakeIcons.adamantArrowhead),
    new Keepsake('Overflowing Cup', 'Dionysus', keepsakeIcons.overflowingCup),
    new Keepsake('Lambent Plume', 'Hermes', keepsakeIcons.lambentPlume),
    new Keepsake('Frostbitten Horn', 'Demeter', keepsakeIcons.frostbittenHorn),
    new Keepsake('Cosmic Egg', 'Chaos', keepsakeIcons.cosmicEgg),
    new Keepsake('Shattered Shackle', 'Sisyphus', keepsakeIcons.shatteredShackle),
    new Keepsake('Evergreen Acorn', 'Eurydice', keepsakeIcons.evergreenAcorn),
    new Keepsake('Broken Spearpoint', 'Patroclus', keepsakeIcons.brokenSpearpoint),
    new Keepsake('Pom Blossom', 'Persephone', keepsakeIcons.pomBlossom),
    new Keepsake('Sigil of the Dead', 'Hades', keepsakeIcons.sigilOfTheDead, true),
]);

export const companions = new Companions().addItems([
    new Companion('Battie', 'Megaera'),
    new Companion('Mort', 'Thanatos'),
    new Companion('Rib', 'Skelly'),
    new Companion('Shady', 'Sisyphus'),
    new Companion('Fidi', 'Dusa'),
    new Companion('Antos', 'Achilles'),
]);

export const mirror = new Mirror()
    .addGroup([
        [
            new MirrorTalent('Shadow Presence', [10, 15, 20, 25, 30]),
            new MirrorTalent('Fiery Presence', [10, 15, 20, 25, 30]),
        ], [
            new MirrorTalent('Chthonic Vitality', [10, 20, 40]),
            new MirrorTalent('Dark Regeneration', [30, 60]),
        ]
    ], true)
    .addGroup([
        [
            new MirrorTalent('Death Defiance', [30, 500, 1000]),
            new MirrorTalent('Stubborn Defiance', [600]),
        ],
        [
            new MirrorTalent('Greater Reflex', [50]),
            new MirrorTalent('Ruthless Reflex', [75]),
        ]
    ], true)
    .addGroup([
        [
            new MirrorTalent('Boiling Blood', [10, 30, 50, 70, 90]),
            new MirrorTalent('Abyssal Blood', [20, 40, 60, 80, 100]),
        ],
        [
            new MirrorTalent('Infernal Soul', [20, 80]),
            new MirrorTalent('Stygian Soul', [0, 60, 180]),
        ]
    ])
    .addGroup([
        [
            new MirrorTalent('Deep Pockets', [30, 35, 40, 45, 50, 55, 60, 65, 70, 75]),
            new MirrorTalent('Golden Touch', [70, 90, 110]),
        ],
        [
            new MirrorTalent('Thick Skin', [40, 45, 50, 55, 60, 65, 70, 75, 80, 85]),
            new MirrorTalent('High Confidence', [50, 100, 150, 200, 250]),
        ]
    ])
    .addGroup([
        [
            new MirrorTalent('Privileged Status', [50, 500]),
            new MirrorTalent('Family Favorite', [50, 150]),
        ],
        [
            new MirrorTalent('Olympian Favor', fillArray(40, 50)),
            new MirrorTalent('Dark Foresight', fillArray(10, 150)),
        ]
    ])
    .addGroup([
        [
            new MirrorTalent('Gods\' Pride', fillArray(20, 100)),
            new MirrorTalent('Gods\' Legacy', fillArray(10, 250)),
        ],
        [
            new MirrorTalent('Fated Authority', [500, 750, 1000, 1250, 1500, 1750, 2000, 2250]),
            new MirrorTalent('Fated Persuasion', [1000, 2000, 3000, 4000]),
        ]
    ]);

export const pact = new Pact()
    .addCondition(new PactCondition('Hard Labor', [1, 1, 1, 1, 1]))
    .addCondition(new PactCondition('Lasting Consequences', [1, 1, 1, 1]))
    .addCondition(new PactCondition('Convenience Fee', [1, 1]))
    .addCondition(new PactCondition('Jury Summons', [1, 1, 1]))
    .addCondition(new PactCondition('Extreme Measures', [1, 2, 3, 4], 3))
    .addCondition(new PactCondition('Calisthenics Program', [1, 1]))
    .addCondition(new PactCondition('Benefits Package', [2, 3]))
    .addCondition(new PactCondition('Middle Management', [2]))
    .addCondition(new PactCondition('Underworld Customs', [2]))
    .addCondition(new PactCondition('Forced Overtime', [3, 3]))
    .addCondition(new PactCondition('Heightened Security', [1]))
    .addCondition(new PactCondition('Routine Inspection', [2, 2, 2, 2]))
    .addCondition(new PactCondition('Damage Control', [1, 1]))
    .addCondition(new PactCondition('Approval Process', [2, 3]))
    .addCondition(new PactCondition('Tight Deadline', [1, 2, 3]));

// weapons.unlockAll();
// keepsakes.unlockAll();
// companions.unlockAll();
// mirror.unlockAll();
// pact.unlockAll();

// weapons.maxLevelAll();
// keepsakes.maxLevelAll();
// companions.maxLevelAll();
// companions.maxCodexAll();

interface Data {
    weapons: Weapons;
    keepsakes: Keepsakes;
    companions: Companions;
    mirror: Mirror;
    pact: Pact;
}

export const defaultData: Data = {
    weapons,
    keepsakes,
    companions,
    mirror,
    pact,
};

export const DataContext = React.createContext<[Data, React.Dispatch<React.SetStateAction<Data>>]>([defaultData, {} as React.Dispatch<React.SetStateAction<Data>>]);

//@ts-ignore
// window.data = defaultData;