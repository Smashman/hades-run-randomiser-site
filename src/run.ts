import { Weapon, WeaponAspect, WeaponOptions, Weapons } from './weapons';
import { KeepsakeConfiguration, KeepsakeOptions, Keepsakes } from './keepsakes';
import { Companion, CompanionOptions, Companions } from './companions';
import { Mirror, MirrorConfiguration, MirrorOptions } from './mirror';
import { Pact, PactConfiguration, PactOptions } from './pact';

export interface RunOptions {
    weaponOptions?: WeaponOptions;
    keepsakeOptions?: KeepsakeOptions;
    companionOptions?: CompanionOptions;
    mirrorOptions?: MirrorOptions;
    pactOptions?: PactOptions;
}

export interface Data {
    weapons: Weapons;
    keepsakes: Keepsakes;
    companions: Companions;
    mirror: Mirror;
    pact: Pact;
}

export const defaultRunOptions: RunOptions = {
    weaponOptions: {
        enabled: true,
        randomAspect: true,
    },
    keepsakeOptions: {
        enabled: true,
        ignoreMaxLevel: true,
        changeBetweenBiomes: true,
    },
    companionOptions: {
        enabled: true,
        ignoreMaxCodex: true,
    },
    mirrorOptions: {
        enabled: true,
    },
    pactOptions: {
        enabled: true,
        // heatRange: {
        //     min: 4,
        //     max: 12,
        // },
    }
};

export class Run {
    constructor(public weapon: Weapon, public weaponAspect: WeaponAspect, public keepsake: KeepsakeConfiguration, public companion: Companion, public mirror: MirrorConfiguration, public pact: PactConfiguration) {
    }

    static generateRandomRun(options: RunOptions, data: Data) {
        const runWeapon = options.weaponOptions?.enabled ? data.weapons.getRandom() : null;
        const runWeaponAspect = runWeapon && options.weaponOptions.randomAspect ? runWeapon.getRandomAspect() : null;
        const runKeepsakes = options.keepsakeOptions?.enabled ? data.keepsakes.getRandomKeepsakes(options.keepsakeOptions) : null;
        const runCompanion = options.companionOptions?.enabled ? data.companions.getRandomCompanion(options.companionOptions) : null;
        const runMirror = options.mirrorOptions?.enabled ? data.mirror.getRandom() : null;
        const runPact = options.pactOptions?.enabled ? data.pact.getRandomWithRange(options.pactOptions.heatRange) : null;

        return new Run(
            runWeapon,
            runWeaponAspect,
            runKeepsakes,
            runCompanion,
            runMirror,
            runPact,
        );
    }
}