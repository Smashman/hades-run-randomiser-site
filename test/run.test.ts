import { Run, RunOptions } from '../src/run';
import * as data from '../src/data';
import { Weapon, WeaponAspect } from '../src/weapons';
import { Keepsake, KeepsakeOptions } from '../src/keepsakes';
import { Companion, CompanionOptions } from '../src/companions';
import { MirrorConfiguration, MirrorTalent } from '../src/mirror';
import { HeatRange, Pact, PactCondition, PactConfiguration, PactOptions} from '../src/pact';
import { fakeIcon } from './helpers';
jest.mock('../src/data');

describe('Run class', () => {
    it('should create expected Run class instance', () => {
        const run = new Run(null, null, null, null, null, null);

        expect(run).toMatchSnapshot();
    });

    describe('generateRandomRun', () => {
        afterEach(() => {
            jest.resetAllMocks();
        });

        it('should generate empty random run with no options', () => {
            const runOptions: RunOptions = {};
            const randomRun = Run.generateRandomRun(runOptions);

            expect(randomRun).toMatchSnapshot();
            expect(data.weapons.getRandom).not.toHaveBeenCalled();
            expect(data.keepsakes.getRandomKeepsakes).not.toHaveBeenCalled();
            expect(data.companions.getRandomCompanion).not.toHaveBeenCalled();
            expect(data.mirror.getRandom).not.toHaveBeenCalled();
            expect(data.pact.getRandomWithRange).not.toHaveBeenCalled();
        });

        it('should generate empty random run with enabled false in all options', () => {
            const runOptions: RunOptions = {
                weaponOptions: {
                    enabled: false,
                },
                keepsakeOptions: {
                    enabled: false,
                },
                companionOptions: {
                    enabled: false,
                },
                mirrorOptions: {
                    enabled: false,
                },
                pactOptions: {
                    enabled: false,
                },
            };
            const randomRun = Run.generateRandomRun(runOptions);

            expect(randomRun).toMatchSnapshot();
            expect(data.weapons.getRandom).not.toHaveBeenCalled();
            expect(data.keepsakes.getRandomKeepsakes).not.toHaveBeenCalled();
            expect(data.companions.getRandomCompanion).not.toHaveBeenCalled();
            expect(data.mirror.getRandom).not.toHaveBeenCalled();
            expect(data.pact.getRandomWithRange).not.toHaveBeenCalled();
        });

        it('should generate random run with weapon enabled', () => {
            const runOptions: RunOptions = {
                weaponOptions: {
                    enabled: true,
                },
            };
            const weapon = new Weapon('Skybreaker', 'Mistblade', 'sword');

            (data.weapons.getRandom as jest.MockedFunction<typeof data.weapons.getRandom>).mockReturnValue(weapon);

            const randomRun = Run.generateRandomRun(runOptions);

            expect(randomRun.weapon).toBe(weapon);

            expect(data.weapons.getRandom).toHaveBeenCalledTimes(1);
            expect(data.keepsakes.getRandomKeepsakes).not.toHaveBeenCalled();
            expect(data.companions.getRandomCompanion).not.toHaveBeenCalled();
            expect(data.mirror.getRandom).not.toHaveBeenCalled();
            expect(data.pact.getRandomWithRange).not.toHaveBeenCalled();
        });

        it('should generate random run with weapon enabled and randomAspect', () => {
            const runOptions: RunOptions = {
                weaponOptions: {
                    enabled: true,
                    randomAspect: true,
                },
            };
            const weaponAspect = new WeaponAspect('Tarak', fakeIcon);
            const weapon = new Weapon('Skybreaker', 'Mistblade', 'sword').addAspect(weaponAspect);

            (data.weapons.getRandom as jest.MockedFunction<typeof data.weapons.getRandom>).mockReturnValue(weapon);

            const getRandomAspectMock = jest.spyOn(weapon, 'getRandomAspect').mockReturnValue(weaponAspect);

            const randomRun = Run.generateRandomRun(runOptions);

            expect(randomRun.weapon).toBe(weapon);
            expect(randomRun.weaponAspect).toBe(weaponAspect);

            expect(data.weapons.getRandom).toHaveBeenCalledTimes(1);
            expect(getRandomAspectMock).toHaveBeenCalledTimes(1);
            expect(data.keepsakes.getRandomKeepsakes).not.toHaveBeenCalled();
            expect(data.companions.getRandomCompanion).not.toHaveBeenCalled();
            expect(data.mirror.getRandom).not.toHaveBeenCalled();
            expect(data.pact.getRandomWithRange).not.toHaveBeenCalled();
        });

        it('should generate random run with keepsake enabled', () => {
            const keepsakeOptions: KeepsakeOptions = {
                enabled: true,
            };
            const runOptions: RunOptions = {
                keepsakeOptions
            };
            const keepsake = new Keepsake('Shoe', 'Jeff', fakeIcon);

            (data.keepsakes.getRandomKeepsakes as jest.MockedFunction<typeof data.keepsakes.getRandomKeepsakes>).mockReturnValue([keepsake]);

            const randomRun = Run.generateRandomRun(runOptions);

            expect(randomRun.keepsake).toStrictEqual([keepsake]);

            expect(data.weapons.getRandom).not.toHaveBeenCalled();
            expect(data.keepsakes.getRandom).not.toHaveBeenCalled();
            expect(data.keepsakes.getRandomKeepsakes).toHaveBeenCalledTimes(1);
            expect(data.keepsakes.getRandomKeepsakes).toHaveBeenCalledWith(keepsakeOptions);
            expect(data.companions.getRandomCompanion).not.toHaveBeenCalled();
            expect(data.mirror.getRandom).not.toHaveBeenCalled();
            expect(data.pact.getRandomWithRange).not.toHaveBeenCalled();
        });

        it('should generate random run with companion enabled', () => {
            const companionOptions: CompanionOptions = {
                enabled: true,
            }
            const runOptions: RunOptions = {
                companionOptions,
            };
            const companion = new Companion('Fluff', 'Jeff');

            (data.companions.getRandomCompanion as jest.MockedFunction<typeof data.companions.getRandomCompanion>).mockReturnValue(companion);

            const randomRun = Run.generateRandomRun(runOptions);

            expect(randomRun.companion).toBe(companion);

            expect(data.weapons.getRandom).not.toHaveBeenCalled();
            expect(data.keepsakes.getRandomKeepsakes).not.toHaveBeenCalled();
            expect(data.companions.getRandom).not.toHaveBeenCalled();
            expect(data.companions.getRandomCompanion).toHaveBeenCalledTimes(1);
            expect(data.companions.getRandomCompanion).toHaveBeenCalledWith(companionOptions);
            expect(data.mirror.getRandom).not.toHaveBeenCalled();
            expect(data.pact.getRandomWithRange).not.toHaveBeenCalled();
        });

        it('should generate random run with mirror enabled', () => {
            const runOptions: RunOptions = {
                mirrorOptions: {
                    enabled: true,
                },
            };
            const mirrorConfiguration = new MirrorConfiguration([new MirrorTalent('Run Fast', [1, 2, 3]), new MirrorTalent('Talk Good', [5, 5, 5, 5])]);

            (data.mirror.getRandom as jest.MockedFunction<typeof data.mirror.getRandom>).mockReturnValue(mirrorConfiguration);

            const randomRun = Run.generateRandomRun(runOptions);

            expect(randomRun.mirror).toStrictEqual(mirrorConfiguration);

            expect(data.weapons.getRandom).not.toHaveBeenCalled();
            expect(data.keepsakes.getRandomKeepsakes).not.toHaveBeenCalled();
            expect(data.companions.getRandomCompanion).not.toHaveBeenCalled();
            expect(data.mirror.getRandom).toHaveBeenCalledTimes(1);
            expect(data.pact.getRandomWithRange).not.toHaveBeenCalled();
        });

        it('should generate random run with pact enabled', () => {
            const pactOptions: PactOptions = {
                enabled: true,
            };
            const runOptions: RunOptions = {
                pactOptions,
            };
            const pact = new Pact().addCondition(new PactCondition('Extra Angry', [1, 1, 1, 2, 3])).addCondition(new PactCondition('Bigger Slam', [2, 2]));
            const pactConfig = new PactConfiguration(pact);

            (data.pact.getRandomWithRange as jest.MockedFunction<typeof data.pact.getRandomWithRange>).mockReturnValue(pactConfig);

            const randomRun = Run.generateRandomRun(runOptions);

            expect(randomRun.pact).toStrictEqual(pactConfig);

            expect(data.weapons.getRandom).not.toHaveBeenCalled();
            expect(data.keepsakes.getRandomKeepsakes).not.toHaveBeenCalled();
            expect(data.companions.getRandomCompanion).not.toHaveBeenCalled();
            expect(data.mirror.getRandom).not.toHaveBeenCalled();
            expect(data.pact.getRandomWithRange).toHaveBeenCalledTimes(1);
            expect(data.pact.getRandomWithRange).toHaveBeenCalledWith(undefined);
        });

        it('should generate random run with pact enabled and a heatRange', () => {
            const heatRange: HeatRange = {
                min: 2,
                max: 12,
            };
            const pactOptions: PactOptions = {
                enabled: true,
                heatRange,
            };
            const runOptions: RunOptions = {
                pactOptions,
            };
            const pact = new Pact().addCondition(new PactCondition('Extra Angry', [1, 1, 1, 2, 3])).addCondition(new PactCondition('Bigger Slam', [2, 2]));
            const pactConfig = new PactConfiguration(pact);

            (data.pact.getRandomWithRange as jest.MockedFunction<typeof data.pact.getRandomWithRange>).mockReturnValue(pactConfig);

            const randomRun = Run.generateRandomRun(runOptions);

            expect(randomRun.pact).toStrictEqual(pactConfig);

            expect(data.weapons.getRandom).not.toHaveBeenCalled();
            expect(data.keepsakes.getRandomKeepsakes).not.toHaveBeenCalled();
            expect(data.companions.getRandomCompanion).not.toHaveBeenCalled();
            expect(data.mirror.getRandom).not.toHaveBeenCalled();
            expect(data.pact.getRandomWithRange).toHaveBeenCalledTimes(1);
            expect(data.pact.getRandomWithRange).toHaveBeenCalledWith(heatRange);
        });
    });
});