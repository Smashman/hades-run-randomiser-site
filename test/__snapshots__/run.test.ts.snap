// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Run class generateRandomRun should generate empty random run with enabled false in all options 1`] = `
Run {
  "companion": null,
  "keepsake": null,
  "mirror": null,
  "pact": null,
  "weapon": null,
  "weaponAspect": null,
}
`;

exports[`Run class generateRandomRun should generate empty random run with no options 1`] = `
Run {
  "companion": null,
  "keepsake": null,
  "mirror": null,
  "pact": null,
  "weapon": null,
  "weaponAspect": null,
}
`;

exports[`Run class should create expected Run class instance 1`] = `
Run {
  "companion": null,
  "keepsake": null,
  "mirror": null,
  "pact": null,
  "weapon": null,
  "weaponAspect": null,
}
`;
